File: backend/src/main/java/de/interaapps/pastefy/auth/strategies/oauth2/providers/GitHubOAuth2Provider.java
package de.interaapps.pastefy.auth.strategies.oauth2.providers;

import org.javawebstack.abstractdata.AbstractObject;
import org.javawebstack.abstractdata.util.QueryString;
import org.javawebstack.http.router.util.MimeType;
import org.javawebstack.httpclient.HTTPClient;
import de.interaapps.pastefy.auth.strategies.oauth2.OAuth2Profile;
import de.interaapps.pastefy.auth.strategies.oauth2.OAuth2Provider;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

public class GitHubOAuth2Provider implements OAuth2Provider {
    private String clientId;
    private String secret;
    private String[] scopes = {"read:user","user:email"};
    public GitHubOAuth2Provider(String clientId, String secret){
        this.clientId = clientId;
        this.secret = secret;
    }

    public GitHubOAuth2Provider setScopes(String... scopes) {
        this.scopes = scopes;
        return this;
    }

    public OAuth2Callback callback(AbstractObject queryParameters, String callbackUrl) {
        AbstractObject abstractObject = new HTTPClient("https://github.com").post("/login/oauth/access_token")
                .formBodyString(new QueryString()
                        .set("client_id", clientId)
                        .set("client_secret", secret)
                        .set("code", queryParameters.string("code"))
                        .toString())
                .header("Accept", MimeType.JSON.getMimeTypes().get(0))
                .data().object();


        if (abstractObject.has("scope")/* && abstractObject.get("scope").string().equals("read:user,user:email")*/) {
            return new OAuth2Callback(this, abstractObject.string("access_token"), abstractObject.has("refresh_token") ? abstractObject.string("refresh_token") : null);
        }

        return null;
    }

    public String redirect(String callbackUrl) {
        try {
            return "https://github.com/login/oauth/authorize?client_id="+clientId+"&scope="+ URLEncoder.encode(String.join(" ", scopes), "UTF-8")+"&redirect_uri="+URLEncoder.encode(callbackUrl, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return "";
    }

    public static class OAuth2Callback extends de.interaapps.pastefy.auth.strategies.oauth2.OAuth2Callback {
        public OAuth2Callback(OAuth2Provider provider, String accessToken, String refreshToken) {
            super(provider, accessToken, refreshToken);
        }

        public HTTPClient createApiClient() {
            return new HTTPClient("https://api.github.com").authorization("token", accessToken);
        }

        public OAuth2Profile getProfile() {
            HTTPClient apiClient = createApiClient();
            OAuth2Profile profile = new OAuth2Profile();

            AbstractObject userData = apiClient.get("/user")
                    .header("Authorization", "token "+accessToken)
                    .data().object();

            if (userData.has("id"))
                profile.id = userData.get("id").number().toString();
            if (userData.has("name"))
                profile.name = userData.get("name").string();
            if (userData.has("avatar_url"))
                profile.avatar = userData.get("avatar_url").string();

            userData.forEach(profile::set);

            apiClient.get("/user/emails")
                    .authorization("token", accessToken)
                    .data().array().forEach(abstractElement -> {
                        if (profile.mail == null) {
                            profile.mail = abstractElement.object().get("email").string();
                        }
                    });
            return profile;
        }
    }
}

File: backend-kt/src/main/kotlin/de/interaapps/pastefy/auth/oauth/providers/GitHubOAuth2Provider.kt
package de.interaapps.pastefy.auth.oauth.providers

import de.interaapps.pastefy.auth.oauth.*

class GitHubOAuth2Provider(
    private val clientId: String = "",
    private val clientSecret: String = "",
    private val scopes: List<String> = listOf("read:user", "user:email"),
    private val http: OAuthHttpClient? = null,
) : OAuth2Provider {
    override val name = "github"

    override fun authorizationUrl(callbackUrl: String, state: String): String = authorizationUrl(
        "https://github.com/login/oauth/authorize",
        mapOf(
            "client_id" to clientId,
            "scope" to scopes.joinToString(" "),
            "redirect_uri" to callbackUrl,
            "state" to state
        ),
    )

    override fun exchangeCode(code: String, callbackUrl: String): OAuth2Tokens = requireHttp().postForm(
        "https://github.com/login/oauth/access_token",
        mapOf("client_id" to clientId, "client_secret" to clientSecret, "code" to code, "redirect_uri" to callbackUrl),
        mapOf("Accept" to "application/json"),
    ).tokens()

    override fun loadProfile(tokens: OAuth2Tokens): OAuth2Profile {
        val headers =
            mapOf("Authorization" to "Bearer ${tokens.accessToken}", "Accept" to "application/vnd.github+json")
        val user = requireHttp().get("https://api.github.com/user", headers)
        val emails = if (user.optionalText("email") == null) requireHttp().get(
            "https://api.github.com/user/emails",
            headers
        ) else null
        val email = user.optionalText("email")
            ?: emails?.firstOrNull { it.path("primary").asBoolean(false) }?.optionalText("email")
            ?: emails?.firstOrNull()?.optionalText("email")
        return OAuth2Profile(
            user.requiredText("id"),
            user.optionalText("name") ?: user.requiredText("login"),
            email,
            user.optionalText("avatar_url")
        )
    }

    private fun requireHttp() = requireNotNull(http) { "OAuth HTTP client is required" }
}
